Add option for using a naitive upload when relaying from discord

yousefmansy1 opened the pull request:
Add option for using a native upload when relaying from discord.
Resolves #1965

```toml
[discord]
	[discord.maps]
		Server="xxxxxxxxxxxxxxxxxxxxxxxxxx"
		Token="xxxxxxxxxxxxxxxxxxxxxxxxxx"
		RemoteNickFormat="{NICK}\n"
		AutoWebhooks=true
		UseLocalAvatar=["telegram.bot"]
		PreserveThreading=true
		UseNativeUpload=true # ! new
```

![image](https://user-images.githubusercontent.com/36427684/219942715-b7e1833b-ec8c-4662-9e39-cba5786742c0.png)

qaisjp commented on bridge/config/config.go:
```diff
@@ -166,6 +166,7 @@ type Protocol struct {
 	URL                    string     // mattermost, slack // DEPRECATED
 	UseAPI                 bool       // mattermost, slack
 	UseLocalAvatar         []string   // discord
+	UseNativeUpload        bool       // discord
```
I think this behaviour should be non-configurable (i.e. just the default behaviour) for any bridges where it's possible. I can't think of a good use-cases for some people wanting external links, and others _not_ wanting external links.

qaisjp replied:
(We can always add a setting later on if someone really wants it. Otherwise I don't think it's worth supporting an extra code path here.)

yousefmansy1 replied:
The reason why I had it configurable is because according to the documentation in the wiki a public link is preferable (referenced in the my issue #1965). 

For most platforms this actually isn't possible (without using the media server)
Discord is unique in this regard.

So we'd be betraying that "standard" by making it non configurable.

Personally, I'm fine with making it the default discord behavior but it's up to the maintainers/community if that's the route we take.

qaisjp replied:
my bad, i missed that part of the documentation. i will post a comment on your issue to help inform the spec for this pull request

yousefmansy1 replied:
So I updated the CR to not make it an option, should drop the commit for that change?

yousefmansy1 replied:
> So I updated the CR to not make it an option, should drop the commit for that change?

@42wim what do you think?

yousefmansy1 commented:
@42wim I think I fixed the `golangci-lint` issues, also squashed the commits into 1.

I removed the error return from the handle download because in reality there isn't really much we can do about it other than log it.

oizyumov commented:
@42wim could you please look into this PR?

Seems like a little lines of code, yet a worthy update to a matterbridge :)

jtrees commented:
Thanks! This is really cool. I've applied the patch and it seems to work... kind of.

The image gets rendered correctly in Element Web and Element Android but not in Cinny. There it just looks like a regular downloadable attachment.

I presume the issue is a missing `mimetype` field. A working image looks like this for me:

```json
{
  "content": {
    "body": "foo.jpeg",
    "info": {
      "h": 512,
      "mimetype": "image/jpeg",
      "size": 1337,
      "w": 512,
      "xyz.amorgan.blurhash": "..."
    },
    "msgtype": "m.image",
    "url": "mxc://matrix.org/blablabla"
  },
  // ...
}
```

The message from Matterbridge looks like this though:

```json
{
  "content": {
    "body": "foo.jpeg",
    "info": {
      "thumbnail_info": {}
    },
    "msgtype": "m.image",
    "url": "mxc://matrix.org/blablabla"
  },
  // ...
}
```

Larslol140 commented on bridge/discord/handlers.go:
```diff
@@ -279,3 +287,14 @@ func handleEmbed(embed *discordgo.MessageEmbed) string {
+		if err != nil {
+			b.Log.Errorf("download %s failed %#v", attach.URL, err)
+		}
+
```
I tested the pull request and had the issue that images were not being displayed on Element web. The image name still contained the parameters of the image url. I'd maybe add code to remove the parameters from an url before using it as filename.

```go

		urlClean, _, _ := strings.Cut(attach.URL, "?")

		helper.HandleDownloadData(b.Log, rmsg, path.Base(urlClean), rmsg.Text, attach.URL, data, b.General)
```

wyf9 commented:
sorry i don't know much about Go language, and i met problems when applying patch:
![image](https://github.com/user-attachments/assets/d61a8ddb-aced-49c8-87f1-1f563734b08e)
*i bridged a Discord server and a Telegram group*
here are some logs:
<details>
<summary>click here to expand logs</summary>

```
[0004]  INFO main:         [main:matterbridge/matterbridge.go:68] Gateway(s) started successfully. Now relaying messages
[0035] DEBUG discord:      [messageCreate:matterbridge/bridge/discord/handlers.go:112] == Receiving event &discordgo.Message{ID:"1362829916931887126", ChannelID:"1125358503305949265", GuildID:"1061629481267245086", Content:"an test image (to get logs)", Timestamp:time.Date(2025, time.April, 18, 16, 39, 32, 352000000, time.Location("")), EditedTimestamp:<nil>, MentionRoles:[]string{}, TTS:false, MentionEveryone:false, Author:(*discordgo.User)(0xc000788340), Attachments:[]*discordgo.MessageAttachment{(*discordgo.MessageAttachment)(0xc000550cb0)}, Components:[]discordgo.MessageComponent{}, Embeds:[]*discordgo.MessageEmbed{}, Mentions:[]*discordgo.User{}, Reactions:[]*discordgo.MessageReactions(nil), Pinned:false, Type:0, WebhookID:"", Member:(*discordgo.Member)(0xc0002d94a0), MentionChannels:[]*discordgo.Channel(nil), Activity:(*discordgo.MessageActivity)(nil), Application:(*discordgo.MessageApplication)(nil), MessageReference:(*discordgo.MessageReference)(nil), ReferencedMessage:(*discordgo.Message)(nil), Interaction:(*discordgo.MessageInteraction)(nil), Flags:0, Thread:(*discordgo.Channel)(nil), StickerItems:[]*discordgo.StickerItem(nil)}
[0035] DEBUG discord:      [HandleDownloadData2:matterbridge/bridge/helper/helper.go:183] Download OK "image.png?ex=6803d1c4&is=68028044&hm=cc9c90f3d1148919ce32d81bce1e2ba27ddeac2157bbf57f897795452b1c7456&" 1383
[0035] DEBUG discord:      [messageCreate:matterbridge/bridge/discord/handlers.go:169] <= Sending message from wyf9 on discord.wyf9server to gateway
[0035] DEBUG discord:      [messageCreate:matterbridge/bridge/discord/handlers.go:170] <= Message is config.Message{Text:"an test image (to get logs)", Channel:"global-chat", Username:"wyf9", UserID:"992995849946804304", Avatar:"https://cdn.discordapp.com/avatars/992995849946804304/a665f2f353733347e37f07c85c4bf52a.jpg", Account:"discord.wyf9server", Event:"", Protocol:"", Gateway:"", ParentID:"", Timestamp:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), ID:"1362829916931887126", Extra:map[string][]interface {}{"file":[]interface {}{config.FileInfo{Name:"image.png?ex=6803d1c4&is=68028044&hm=cc9c90f3d1148919ce32d81bce1e2ba27ddeac2157bbf57f897795452b1c7456&", Data:(*[]uint8)(0xc0006c3a40), Comment:"an test image (to get logs)", URL:"https://cdn.discordapp.com/attachments/1125358503305949265/1362829916697133277/image.png?ex=6803d1c4&is=68028044&hm=cc9c90f3d1148919ce32d81bce1e2ba27ddeac2157bbf57f897795452b1c7456&", Size:0, Avatar:false, SHA:"", NativeID:""}}}}
[0035] DEBUG gateway:      [SendMessage:matterbridge/gateway/gateway.go:500] => Sending config.Message{Text:"an test image (to get logs)", Channel:"global-chat", Username:"wyf9", UserID:"992995849946804304", Avatar:"https://cdn.discordapp.com/avatars/992995849946804304/a665f2f353733347e37f07c85c4bf52a.jpg", Account:"discord.wyf9server", Event:"", Protocol:"discord", Gateway:"global-chat", ParentID:"", Timestamp:time.Date(2025, time.April, 19, 0, 39, 32, 901589026, time.Local), ID:"1362829916931887126", Extra:map[string][]interface {}{"file":[]interface {}{config.FileInfo{Name:"image.png?ex=6803d1c4&is=68028044&hm=cc9c90f3d1148919ce32d81bce1e2ba27ddeac2157bbf57f897795452b1c7456&", Data:(*[]uint8)(0xc0006c3a40), Comment:"an test image (to get logs)", URL:"https://cdn.discordapp.com/attachments/1125358503305949265/1362829916697133277/image.png?ex=6803d1c4&is=68028044&hm=cc9c90f3d1148919ce32d81bce1e2ba27ddeac2157bbf57f897795452b1c7456&", Size:0, Avatar:false, SHA:"", NativeID:""}}}} from discord.wyf9server (global-chat) to api.tailchat (api)
[0035] DEBUG api:          [Send:matterbridge/bridge/api/api.go:117] enqueueing message from [**DC**] @wyf9 :  on ring buffer
[0035] DEBUG gateway:      [func1:matterbridge/gateway/gateway.go:509] => Send from discord.wyf9server (global-chat) to api.tailchat (api) took 532.821µs
[0035] DEBUG gateway:      [SendMessage:matterbridge/gateway/gateway.go:500] => Sending config.Message{Text:"an test image (to get logs)", Channel:"global-chat", Username:"wyf9", UserID:"992995849946804304", Avatar:"https://cdn.discordapp.com/avatars/992995849946804304/a665f2f353733347e37f07c85c4bf52a.jpg", Account:"discord.wyf9server", Event:"", Protocol:"discord", Gateway:"global-chat", ParentID:"", Timestamp:time.Date(2025, time.April, 19, 0, 39, 32, 901589026, time.Local), ID:"1362829916931887126", Extra:map[string][]interface {}{"file":[]interface {}{config.FileInfo{Name:"image.png?ex=6803d1c4&is=68028044&hm=cc9c90f3d1148919ce32d81bce1e2ba27ddeac2157bbf57f897795452b1c7456&", Data:(*[]uint8)(0xc0006c3a40), Comment:"an test image (to get logs)", URL:"https://cdn.discordapp.com/attachments/1125358503305949265/1362829916697133277/image.png?ex=6803d1c4&is=68028044&hm=cc9c90f3d1148919ce32d81bce1e2ba27ddeac2157bbf57f897795452b1c7456&", Size:0, Avatar:false, SHA:"", NativeID:""}}}} from discord.wyf9server (global-chat) to telegram.siiwaychatbridge (-1002206757362)
[0035] DEBUG telegram:     [Send:matterbridge/bridge/telegram/telegram.go:121] => Receiving config.Message{Text:"an test image (to get logs)", Channel:"-1002206757362", Username:"[<b>DC</b>] @wyf9 : ", UserID:"992995849946804304", Avatar:"https://cdn.discordapp.com/avatars/992995849946804304/a665f2f353733347e37f07c85c4bf52a.jpg", Account:"discord.wyf9server", Event:"", Protocol:"discord", Gateway:"global-chat", ParentID:"", Timestamp:time.Date(2025, time.April, 19, 0, 39, 32, 901589026, time.Local), ID:"", Extra:map[string][]interface {}{"file":[]interface {}{config.FileInfo{Name:"image.png?ex=6803d1c4&is=68028044&hm=cc9c90f3d1148919ce32d81bce1e2ba27ddeac2157bbf57f897795452b1c7456&", Data:(*[]uint8)(0xc0006c3a40), Comment:"an test image (to get logs)", URL:"https://cdn.discordapp.com/attachments/1125358503305949265/1362829916697133277/image.png?ex=6803d1c4&is=68028044&hm=cc9c90f3d1148919ce32d81bce1e2ba27ddeac2157bbf57f897795452b1c7456&", Size:0, Avatar:false, SHA:"", NativeID:""}}}}
[0035] DEBUG telegram:     [TGGetParseMode:matterbridge/bridge/telegram/telegram.go:70] Using mode HTML
[0036] DEBUG gateway:      [SendMessage:matterbridge/gateway/gateway.go:519] mID telegram.siiwaychatbridge: 653
[0036] DEBUG gateway:      [func1:matterbridge/gateway/gateway.go:509] => Send from discord.wyf9server (global-chat) to telegram.siiwaychatbridge (-1002206757362) took 798.503478ms
```

</details>
